src/red_black_tree.jl

```diff
@@ -0,0 +1,199 @@
+		else
+			root.leftChild = insert(root.leftChild, node)
+			root.leftChild.parent = root
+		end
+		return root
+	end
+end
+
+function left_rotate!(tree::RBTree, node_x::RBTreeNode)
+	node_y = node_x.rightChild
+	node_x.rightChild = node_y.leftChild
+	if node_x.rightChild != nothing
+		node_x.rightChild.parent = node_x
+	end
+	node_y.parent = node_x.parent
+	if (node_x.parent == nothing)
+		tree.root = node_y
+	elseif (node_x == node_x.parent.leftChild)
+		node_x.parent.leftChild = node_y
+	else
+		node_x.parent.rightChild = node_y
+	end
+	node_y.leftChild = node_x
+	node_x.parent = node_y
+end	
+
+function right_rotate!(tree::RBTree, node_x::RBTreeNode)
+	node_y = node_x.leftChild
+	node_x.leftChild = node_y.rightChild
+	if node_x.leftChild != nothing
+		node_x.leftChild.parent = node_x
+	end
+	node_y.parent = node_x.parent
+	if (node_x.parent == nothing)
+		tree.root = node_y
+	elseif (node_x == node_x.parent.leftChild)
+		node_x.parent.leftChild = node_y
+	else
+		node_x.parent.rightChild = node_y
+	end
+	node_y.rightChild = node_x
+	node_x.parent = node_y
+end	
+
+node_color(color::Bool) = (color) ? "RED" : "BLACK"
+desc_node(relation, node::RBTreeNode) = println("$relation is ", node.data, " color is ", node_color(node.color))
+
+function balance!(tree::RBTree, node::RBTreeNode)
+	parent = nothing
+	grand_parent = nothing
+	# for root node, we need to change the color to black
+	# other nodes, we need to maintain the property such that
+	# no two adjacent nodes are red in color
+	while (node != tree.root && node.color && node.parent.color)
+		parent = node.parent
+		grand_parent = parent.parent
+		
+		# parent is the leftChild of grand_parent
+		if (parent == grand_parent.leftChild)
+			uncle = grand_parent.rightChild
+
+			# uncle is red in color
+			if (uncle != nothing && uncle.color)
+				grand_parent.color = true
+				parent.color = false
+				uncle.color = false
+				node = grand_parent
+			# uncle is black in color
+			else 
+				# node is rightChild of it's parent
+				if (node == parent.rightChild)
+					left_rotate!(tree, parent)
+					node = parent
+					parent = node.parent
+				end
+				# node is leftChild of it's parent
+				right_rotate!(tree, grand_parent)
+				parent.color, grand_parent.color = grand_parent.color, parent.color
+				node = parent
+			end
+		# parent is the rightChild of grand_parent
+		else 
+			uncle = grand_parent.leftChild
+
+			# uncle is red in color
+			if (uncle != nothing && uncle.color)
+				grand_parent.color = true
+				parent.color = false
+				uncle.color = false
+				node = grand_parent
+			# uncle is black in color
+			else 
+				# node is leftChild of it's parent
+				if (node == parent.leftChild)
+					right_rotate!(tree, parent)
+					node = parent
+					parent = node.parent
+				end
+				# node is rightChild of it's parent
+				left_rotate!(tree, grand_parent)
+				parent.color, grand_parent.color = grand_parent.color, parent.color
+				node = parent
+			end
+		end
+	end
```

Comment:
this file is using tabs instead of spaces.

Reply:
resolved in the new commit